synapse/lib/layer.py

```diff
@@ -4093,6 +4093,104 @@ async def iterNodeDataKeys(self, buid):
             prop = self.getAbrvProp(abrv)
             yield prop[0]
 
+    async def iterLayerAddPerms(self):
+
+        # nodes & props
+        for byts, abrv in self.propabrv.slab.scanByFull(db=self.propabrv.name2abrv):
+            form, prop = s_msgpack.un(byts)
+            if form is None:
+                continue
+
+            if self.layrslab.prefexists(abrv, db=self.byprop):
+                if prop:
+                    yield ('node', 'prop', 'set', f'{form}:{prop}')
+                else:
+                    yield ('node', 'add', form)
+
+            await asyncio.sleep(0)
+
+        # tagprops
+        for byts, abrv in self.tagpropabrv.slab.scanByFull(db=self.tagpropabrv.name2abrv):
+            info = s_msgpack.un(byts)
+            if None in info or len(info) != 3:
+                continue
+
+            if self.layrslab.prefexists(abrv, db=self.bytagprop):
+                yield ('node', 'tag', 'add', *info[1].split('.'))
+
+            await asyncio.sleep(0)
+
+        # nodedata
+        for abrv in self.dataslab.scanKeys(db=self.dataname):
+            name, _ = self.getAbrvProp(abrv)
+            yield ('node', 'data', 'set', name)
+
+            await asyncio.sleep(0)
+
+        # edges
+        for verb in self.layrslab.scanKeys(db=self.byverb):
+            yield ('node', 'edge', 'add', verb.decode())
+
+            await asyncio.sleep(0)
+
+        # tags
+        # NB: tag perms should be yielded for every leaf on every node in the layer
+        async with self.core.getSpooledDict() as tags:
+
+            # Collect all tag abrvs for all nodes in the layer
+            for lkey, buid in self.layrslab.scanByFull(db=self.bytag):
+                abrv = lkey[:8]
+                abrvs = list(tags.get(buid, []))
+                abrvs.append(abrv)
+                await tags.set(buid, abrvs)
+
+                await asyncio.sleep(0)
```

Comment:
I think for this loop at least, sleeping every iteration is a bit much and we could limit it to every 1000 iterations or something, I did a quick perf test and there is a notable impact here.